### DISCUSSION TITLE
How do I embed data in a Quarto blog?

### DISCUSSION THREAD

**Discussion by pgseye** [quarto-dev/quarto-cli#7717] [Q&A] [answered]
### Description

I recently made a blog with Quarto that works well for my needs. Until now I've just used toy data but I would like to be able to include a larger dataset. I want the reader to be able to use the embedded data to run the code for themselves, without me linking to it from my Dropbox or something similar.

I read about `Observable` and `FileAttachment`, but don't really understand its application. What is the difference to simply including the e.g. csv file in a sub-folder and using read.csv. I don't understand the difference in the behaviours. e.g Something like:

```{ojs}
dat = FileAttachment("data/heart.csv").csv({typed: true})
```

```{r}
dat <- read.csv("data/heart.csv", header = T)
```

I've only tested this locally and the latter works. I don't know how to test this live without uploading a draft (and if I include draft = true in the YAML I can't seem to preview it anyway)
When I use `FileAttachment` the data displays as a list, not a data frame.

Thanks.

**Comment by mcanouil** [comment]
Code is evaluated at rendering, only the parts that is part of the output will remain.

Observable is evaluated after rendering and in the browser.

So, the former only needs the csv locally when rendering while the latter will need it everytime you open the HTML.

I suggest for you to get acquainted with what Observable.js is compared to R/Python/etc.
Possibly through the documentation inside Quarto which has links to other good resources, especially <https://quarto.org/docs/interactive/ojs/data-sources.html>.

**Reply by pgseye** [reply]
Thanks,
I'd read that previously and it still wasn't clear to me based on the example. All I am wanting to know is how a reader of my page can copy the code chunk and automatically load the linked data. The only other (clunky) option I can seem to see is to use datatable to display and allow downloading of the data frame as a .csv/.xlsx/etc, but then one would have to take extra steps to import that back into their R session. Maybe what I am asking is not possible.

**Reply by mcanouil** [reply, marked as answer]
You cannot copy data from file with neither approaches.
This is similar as in just using an R script reading data. If you don't share the data alongside the script, the script won't work, no matter you copy/paste the content of the R script.
Make a link to download load the data hosted where you want `[Download the data](path/or/url/to/file.csv)`.

**Reply by pgseye** [reply]
Thank you
